docs/sources.md

```diff
@@ -0,0 +1,268 @@
+as the value of `id`.
+
+The returned value from `Resolve()` must implement [SourceInstance](../source/manager.go#L18):
+
+```go
+type SourceInstance interface {
+	CacheKey(ctx context.Context, g session.Group, index int) (string, string, solver.CacheOpts, bool, error)
+	Snapshot(ctx context.Context, g session.Group) (cache.ImmutableRef, error)
+}
+```
+
+A `SourceInstance` is returned by `Resolve()`, so it is assumed already to know what its identifier is, e.g. `docker.io/library/nginx:latest`.
+
+The two functions in the signature of the `SourceInstance` are expected to perform the following functions.
+
+#### `CacheKey`
+
+`CacheKey()` gives the solver a key that uniquely identifies the instance of the operation (including its arguments and inputs).
+So for a container image source, it would basically amount to the image digest (plus some other metadata).
+
+The purpose of `CacheKey()` is for the solver to be able to ask the source, "I have this `id`, what unique cache key is there for it?"
+If the response is identical to the one already in cache, the solver knows it does not need to retrieve the data again, and simply can
+use what is in its cache. If it is different or does not exist, then it knows it needs to ask the source for the content referred to by the
+`id`
+
+This is expected to be idempotent, a read-only activity that returns a unique cache key.
+
+The arguments to `CacheKey()` are:
+
+* `ctx context.Context`: A golang [context](https://pkg.go.dev/context#Context).
+* `g session.Group`: The [session.Group](https://pkg.go.dev/github.com/moby/buildkit@v0.10.1/session#Group) contains parameters to identify the specific buildkit client calling this build. These may be useful for things like credentials.
+* `index int`: The `CacheKey()` can be called multiple times for the same source item, each indicating that it is a separate subitem. The `index` indicates which iteration of call this is. Generally, if the `index > 0` then it is complete.
+
+The expected return values are:
+
+* `cacheKey string`: the unique cache ID; this might be the combination of a hash with the source identifier. For example, `docker-image` uses the hash of the image config, while `local` combines the path to the directory and hash of a json that includes the session and search parameters.
+* `imgDigest string`: the digest of the root of the image. For example, `docker-image` uses the hash of the root manifest or index, while `local` returns the hash of the constructed json of the session and search parameters.
+* `cacheOpts solver.CacheOpts`: The [CacheOpts](https://pkg.go.dev/github.com/moby/buildkit@v0.10.1/solver#CacheOpts) contain metadata that can be passed from one iteration of `CacheKey()` to another, enabling context to be passed from call to call without the `Source` needing to store information. This is useful for items like progress streams and remote reference information.
```

Comment:
What are examples of some opts here? And who consumes them, given that the `SourceInstance` only returns them, so who consumes them and how?